compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/SceneState.kt

```diff
@@ -0,0 +1,411 @@
+             *   It is public for iOS 17 and hope back ported for iOS 13 as well (but we need to check)
+             */
+            if (sceneView.isReadyToShowContent.value) {
+                EntrypointCompositionLocals {
+                    SceneCompositionLocal {
+                        content()
+                    }
+                }
+            }
+        }
+    }
+
+    private val safeAreaState: MutableState<PlatformInsets> by lazy {
+        //TODO It calc 0,0,0,0 on initialization
+        mutableStateOf(calcSafeArea())
+    }
+    private val layoutMarginsState: MutableState<PlatformInsets> by lazy {
+        //TODO It calc 0,0,0,0 on initialization
+        mutableStateOf(calcLayoutMargin())
+    }
+
+    override fun updateSafeArea() {
+        safeAreaState.value = calcSafeArea()
+        layoutMarginsState.value = calcLayoutMargin()
+    }
+
+    @Composable
+    private fun SceneCompositionLocal(content: @Composable () -> Unit) =
+        CompositionLocalProvider(
+            LocalUIKitInteropContext provides interopContext,
+            LocalKeyboardOverlapHeight provides keyboardVisibilityListener.keyboardOverlapHeightState.value,
+            LocalSafeArea provides safeAreaState.value,
+            LocalLayoutMargins provides layoutMarginsState.value,
+            content = content
+        )
+
+    override fun display(focusable: Boolean, onDisplayed: () -> Unit) {
+        sceneView.onAttachedToWindow = {
+            sceneView.onAttachedToWindow = null
+            updateLayout()
+            onDisplayed()
+            if (focusable) {
+                focusStack.pushAndFocus(sceneView)
+            }
+        }
+        rootViewController.view.addSubview(sceneView)
+    }
+
+    override fun dispose() {
+        if (focusable) {
+            focusStack.popUntilNext(sceneView)
+        }
+        sceneView.dispose()
+        scene.close()
+        // After scene is disposed all UIKit interop actions can't be deferred to be synchronized with rendering
+        // Thus they need to be executed now.
+        interopContext.retrieve().actions.forEach { it.invoke() }
+    }
+
+    override fun needRedraw() = sceneView.needRedraw()
+
+    override fun setLayout(value: SceneLayout) {
+        _layout = value
+        when (value) {
+            SceneLayout.UseConstraintsToFillContainer -> {
+                delegate.metalOffset = Offset.Zero
+                sceneView.setFrame(CGRectZero.readValue())
+                sceneView.translatesAutoresizingMaskIntoConstraints = false
+                constraints = listOf(
+                    sceneView.leftAnchor.constraintEqualToAnchor(rootViewController.view.leftAnchor),
+                    sceneView.rightAnchor.constraintEqualToAnchor(rootViewController.view.rightAnchor),
+                    sceneView.topAnchor.constraintEqualToAnchor(rootViewController.view.topAnchor),
+                    sceneView.bottomAnchor.constraintEqualToAnchor(rootViewController.view.bottomAnchor)
```

Comment:
Another good candidate for extension function

Reply:
Better to leave it here.
Because there are to many arguments.